### PR TITLE
Bug 1824840, Document manual creation of IAM for AWS, GCP, and Azure

### PR DESCRIPTION
**Pull request by ahardin-rh**
https://bugzilla.redhat.com/show_bug.cgi?id=1824840

Preview builds: 

AWS content:
http://file.rdu.redhat.com/~ahardin/May-13-2020/manual-IAM-creation/installing/installing_aws/manually-creating-iam.html

Azure content:
http://file.rdu.redhat.com/~ahardin/May-13-2020/manual-IAM-creation/installing/installing_azure/manually-creating-iam-azure.html

GCP content:
http://file.rdu.redhat.com/~ahardin/May-13-2020/manual-IAM-creation/installing/installing_gcp/manually-creating-iam-gcp.html

### CODE REVIEW

**Comment by ahardin-rh** [comment]
@pweil- @dgoodwin Took my first pass at this. PTAL. Thanks!

**Comment by pweil-** [comment]
Thanks for putting it together!

We'll want to hold merging until we get the jira linked in the bug worked.  Overall looks good to me but we'll need to include info from Devan's work that prevents clusters from upgrading if this is used as well as any tooling we have to help customers curate the permissions and ack that they're ready to upgrade.

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:56]: I think it would be exceedingly rare a customer would be thinking about creating their own release images and a terrible idea for a lot of reasons, probably safe to drop this sentence.

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:58]: If you think it might be better, we could skip the base64 encoding and use stringData instead of data, a convenience Kube makes available to b64encode for you when creating only. (won't be present if you view the yaml/json later) It might simplify the docs a little bit.

**Review by dgoodwin — COMMENTED** [review]
Looks fantastic, just a couple little comments.

However I don't think we can merge this yet as we don't really want customers doing this because we believe there will be problems on upgrades that will block and fail the upgrade with no clear indication to user on what is wrong or how to fix. 

The epic to track disabled CCO is slated for 4.6: https://issues.redhat.com/browse/CO-925

File: installing/installing_aws/manually-creating-iam.adoc
Comment (JoelSpeed) [inline installing/installing_aws/manually-creating-iam.adoc:46]: Should this image be a placeholder or release rather than referring to a CI build?

File: installing/installing_aws/manually-creating-iam.adoc
Comment (sferich888) [inline installing/installing_aws/manually-creating-iam.adoc:49]: This is sufficiently complex, can we simplify this or demonstrate exactly what we want people to do?

**Review by sferich888 — CHANGES_REQUESTED** [review]

**Comment by dgoodwin** [comment]
We are a go for documenting this possibly back as far as 4.3. We're going to be doing some testing to see how it's behaving on other clouds and will be back with some more in-depth comments on this procedure ASAP.

/hold

**Comment by sferich888** [comment]
Are we missing something with this PR? While its good that we can change or modify the IAM user, what about modifying its policy or the policy it uses? 

https://access.redhat.com/solutions/5056391
https://issues.redhat.com/browse/CO-925
https://issues.redhat.com/browse/CO-952

**Comment by dgoodwin** [comment]
I believe this PR is specific to your last two links, and not the first. Master role permissions is not something covered in the scope of this, this is purely about the credentials operator that mints creds for the openshift components such as ingress/machine/registry operator. These credentials should not be adjusted by a user.

Masters receive their creds via instance roles it seems, which is a separate thing.

**Comment by sferich888** [comment]
Yes but my thinking is that users may conflate the two and we may need to look at how we explaine these topics as a cohesive story.

**Comment by dgoodwin** [comment]
I would definitely agree, a cohesive way to present the whole picture spanning installer, master roles, and the cred operator would be really nice.

**Comment by dgoodwin** [comment]
This sprint I have been testing and updating the repo docs for both manual credentials, and removing root creds post-install.

https://github.com/openshift/cloud-credential-operator/blob/master/README.md
https://issues.redhat.com/browse/CO-956 (remove root creds)
https://issues.redhat.com/browse/CO-980 (manual creds)

I've pinged @ahardin-rh about the remove root creds one, but it does feel like there may need to be a larger effort to document all of this as well as the permissions required across installer and master IAM roles.

**Comment by ahardin-rh** [comment]
@dgoodwin I was able to address this PR a bit more today, but feel like it's missing some connective tissue. Perhaps we can sync up to discuss this further after the holiday weekend? I will find time on your calendar. Thanks!

File: modules/mint-mode-with-removal-of-admin-credential.adoc
Comment (dgoodwin) [inline modules/mint-mode-with-removal-of-admin-credential.adoc:10]: Could we clarify that this is an action that must be taken by a human administrator.

File: modules/mint-mode-with-removal-of-admin-credential.adoc
Comment (dgoodwin) [inline modules/mint-mode-with-removal-of-admin-credential.adoc:5]: Is this doc section AWS specific? The feature is only available on AWS right now, though we hope to land it for Azure and GCP soon.

File: installing/installing_aws/installing-aws-default.adoc
Comment (dgoodwin) [inline installing/installing_aws/installing-aws-default.adoc:33]: Suggest: "then a cluster administrator can [manually create and maintain IAM credentials]". I like clarifying it's not just creation but also on-going management across upgrades so folks know what they're getting into.

File: installing/installing_aws/installing-aws-default.adoc
Comment (dgoodwin) [inline installing/installing_aws/installing-aws-default.adoc:35]: For the last sentence, suggest: "Manual mode can also be used in environments where the cloud IAM APIs are not reachable."

Lets avoid any mention of calling the operator "disabled", that was a bad call on my part, we expect in manual mode it will do increasingly more and I don't want people thinking it's actually disabled and then it starts reporting problems they don't want to deal with.

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:2]: Manual mode is supported on Azure and GCP as well.

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:14]: Can you think of a good spot to try to steer people away from this and towards using the admin cred normally? We consider this by far the best practice. Maybe some warnings here?

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:39]: By the time 4.6 launches there'll be a better and new mechanism for doing this using an actual config resource that no longer calls it disabled, rather manual mode. 

Prior to 4.6 however the above configmap will be the only option, so if we backport this for 4.5 (where manual mode does work) we would need to change this portion.

cc @joelddiaz

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:73]: This is available in 4.6 and should be the recommended option there, maybe with a fallback if people are on older releases. The new command is:

oc adm release extract --credentials-requests --cloud=aws|azure|gcp|openstack|ovirt|vsphere

File: installing/installing_aws/manually-creating-iam.adoc
Comment (dgoodwin) [inline installing/installing_aws/manually-creating-iam.adoc:132]: I would drop the last sentence here, that's probably from our repo but not really relevant to users. Not even true actually, we're still trying to get the CI job ready for it.

**Review by dgoodwin — CHANGES_REQUESTED** [review]

File: installing/installing_aws/installing-aws-customizations.adoc
Comment (joelddiaz) [inline installing/installing_aws/installing-aws-customizations.adoc:37]: instead of "cane be disabled", maybe "can be bypassed"?

File: installing/installing_aws/manually-creating-iam.adoc
Comment (joelddiaz) [reply installing/installing_aws/manually-creating-iam.adoc:14]: There's also the option of removing the admin cred post-install. This allows the auto-managed mine-grained creds, but also doesn't leave the admin creds in the cluster after they aren't needed. (This does complicate upgrades if updated credentials are needed and CCO cannot update them w/o the admin creds).
cc @dgoodwin

File: installing/installing_aws/manually-creating-iam.adoc
Comment (joelddiaz) [reply installing/installing_aws/manually-creating-iam.adoc:39]: What @dgoodwin is refering to here is the enhancement proposal ( https://github.com/openshift/enhancements/blob/master/enhancements/installer/aws-permissions-check-bypass.md ) to allow bypassing permissions checks before install. A side-effect of enabling this functionality is that we will also (using the same proposed CCO config resource) be able to force CCO into "manual" mode w/o the need for this configmap.

File: modules/mint-mode-with-removal-of-admin-credential.adoc
Comment (ahardin-rh) [reply modules/mint-mode-with-removal-of-admin-credential.adoc:5]: Yes, I am setting this up so that it only appears in the AWS section.

File: installing/installing_aws/manually-creating-iam.adoc
Comment (ahardin-rh) [reply installing/installing_aws/manually-creating-iam.adoc:39]: @dgoodwin @joelddiaz Once this PR merges with the 4.5 content, we can do a follow-up with 4.6 content if that works for you.

File: installing/installing_aws/manually-creating-iam.adoc
Comment (ahardin-rh) [reply installing/installing_aws/manually-creating-iam.adoc:2]: Restructuring to include Azure and GCP. Thanks!

File: installing/installing_aws/manually-creating-iam.adoc
Comment (ahardin-rh) [reply installing/installing_aws/manually-creating-iam.adoc:73]: Thanks! I'm noting to include this in our follow-up PR for 4.6.

File: installing/installing_aws/installing-aws-customizations.adoc
Comment (ahardin-rh) [reply installing/installing_aws/installing-aws-customizations.adoc]: Updated with language with Devan's suggestion.

**Comment by ahardin-rh** [comment]
@joelddiaz @dgoodwin I took another pass at this and restructured for AWS, Azure, and GCP. Preview builds:

AWS content:
http://file.rdu.redhat.com/~ahardin/May-13-2020/manual-IAM-creation/installing/installing_aws/manually-creating-iam.html

Azure content:
http://file.rdu.redhat.com/~ahardin/May-13-2020/manual-IAM-creation/installing/installing_azure/manually-creating-iam-azure.html

GCP content:
http://file.rdu.redhat.com/~ahardin/May-13-2020/manual-IAM-creation/installing/installing_gcp/manually-creating-iam-gcp.html

Once we get this content reviewed, approved, and merged, we can create a PR with 4.6 updates. 

PTAL. Thanks!

**Comment by joelddiaz** [comment]
@ahardin-rh the content looks good (I reviewed the generated files you provided). I did have some thoughts as I reviewed it.

Each of the linked files are titled "Manually creating IAM - Installing on <cloud name>", but then in each of those documents, it lists information for other clouds. It almost feels like the document - as it is written today - covers all clouds simultaneously, but then it is copy-pasted as-is for each of the clouds we want to document. It is probably fine as it is, but could cause confusion if someone looking through the Azure/GCP version of the doc then sees examples showing AWS details.
Outside of having one document per-cloud (which is quite a bit more documentation so that the examples are cloud-specific), having the document exist outside of any cloud-specific subsection would clear that up (and reduce the duplication of documentation).
Either way, I don't think it's a blocker, but just something I noticed as I reviewed.

**Comment by ahardin-rh** [comment]
@joelddiaz Good point! I will see what I can do as far as conditionalizing content for each cloud provider so that only the pertinent content renders for each.

**Comment by dgoodwin** [comment]
Looks good to me.

**Comment by ahardin-rh** [comment]
@joelddiaz @dgoodwin Can you please provide an equivalent example of the following for GCP and Azure? 

Sample CredentialsRequest:
```
apiVersion: cloudcredential.openshift.io/v1
kind: CredentialsRequest
metadata:
  name: cloud-credential-operator-iam-ro
  namespace: openshift-cloud-credential-operator
spec:
  secretRef:
    name: cloud-credential-operator-iam-ro-creds
    namespace: openshift-cloud-credential-operator
  providerSpec:
    apiVersion: cloudcredential.openshift.io/v1
    kind: AWSProviderSpec
    statementEntries:
    - effect: Allow
      action:
      - iam:GetUser
      - iam:GetUserPolicy
      - iam:ListAccessKeys
      resource: "*"
```

**Comment by ahardin-rh** [comment]
@lwan-wanglin Can you please provide QE review? Thank you!

**Comment by ahardin-rh** [comment]
Once this PR merges, I will create a separate PR for 4.3 to apply this content, but remove Mint + Remove Admin Cred from the AWS section per https://github.com/openshift/cloud-credential-operator#support-matrix

**Comment by joelddiaz** [comment]
> @joelddiaz @dgoodwin Can you please provide an equivalent example of the following for GCP and Azure?

for Azure https://github.com/openshift/cluster-image-registry-operator/blob/master/manifests/01-registry-credentials-request-azure.yaml

for GCP https://github.com/openshift/cluster-image-registry-operator/blob/master/manifests/01-registry-credentials-request-gcs.yaml

**Comment by lwan-wanglin** [comment]
> @lwan-wanglin Can you please provide QE review? Thank you!

@yunjiang29 is QE for this function, I invite him to review.

File: modules/manually-create-identity-access-management.adoc
Comment (yunjiang29) [inline modules/manually-create-identity-access-management.adoc:76]: There will be lots of output here, I'm afraid that user may make mistake when finding the CR text
so IMO, beside above command, it is better that list all cloud provider's CR here for user reference.
@dgoodwin @joelddiaz WDYT?

Comment (dgoodwin) [reply modules/manually-create-identity-access-management.adoc:76]: I would not want to embed all the credentials details in the documentation as that would be very awkward to maintain release by release. 

We have however implemented an oc adm command to simplify this in 4.6 so in very near future we could update these docs to avoid any mistakes in this portion. At present I think this may be the best option for now, users are opting into the possibility of making mistakes in moving to manual credentials management. We hope to make it easier with additional tooling in future.

File: modules/manually-create-identity-access-management.adoc
Comment (adellape) [inline modules/manually-create-identity-access-management.adoc:161]: Maybe "Create YAML files for secrets [...]", so that it doesn't sound like the files themselves are secret. :detective:

File: modules/manually-create-identity-access-management.adoc
Comment (adellape) [inline modules/manually-create-identity-access-management.adoc:162]: "must be stored using the namespace and secret name defined in [...]"

File: modules/mint-mode.adoc
Comment (adellape) [inline modules/mint-mode.adoc:20]: s/needs/requires/

File: modules/admin-credentials-root-secret-formats.adoc
Comment (adellape) [inline modules/admin-credentials-root-secret-formats.adoc:21]: Capitalize "Mint Mode"? Noticed it was done so elsewhere in the doc.

File: modules/mint-mode-with-removal-of-admin-credential.adoc
Comment (adellape) [inline modules/mint-mode-with-removal-of-admin-credential.adoc:11]: Mint Mode capitalized?

File: modules/admin-credentials-root-secret-formats.adoc
Comment (adellape) [inline modules/admin-credentials-root-secret-formats.adoc:29]: Suggest putting "secret format" after the cloud provider name in each of these code block titles, e.g., "Amazon Web Services (AWS) secret format".

File: modules/mint-mode-with-removal-of-admin-credential.adoc
Comment (adellape) [inline modules/mint-mode-with-removal-of-admin-credential.adoc:15]: s/needed/required/

But also maybe break up this sentence if possible?

File: modules/mint-mode-with-removal-of-admin-credential.adoc
Comment (adellape) [inline modules/mint-mode-with-removal-of-admin-credential.adoc:16]: "After the associated credential is removed, it can be destroyed [...]"

File: modules/manually-create-identity-access-management.adoc
Comment (adellape) [inline modules/manually-create-identity-access-management.adoc:50]: Throughout, suggest reassessing whether "admin" should be spelled out "administrator" or marked up `admin`.

File: modules/admin-credentials-root-secret-formats.adoc
Comment (adellape) [inline modules/admin-credentials-root-secret-formats.adoc:72]: Example output block for this line?

File: modules/admin-credentials-root-secret-formats.adoc
Comment (adellape) [inline modules/admin-credentials-root-secret-formats.adoc:77]: Could use `[source,yaml`] here.

**Review by adellape — COMMENTED** [review]
Some suggestions, but nothing critical. :+1:

File: modules/manually-create-identity-access-management.adoc
Comment (yunjiang29) [reply modules/manually-create-identity-access-management.adoc:76]: Right, oc tool is a great option for users.

**Comment by yunjiang29** [comment]
LGTM @ahardin-rh

File: modules/admin-credentials-root-secret-formats.adoc
Comment (ahardin-rh) [inline modules/admin-credentials-root-secret-formats.adoc:39]: @dgoodwin Should we have  Base64encodeAccessKeyID and similar values in each of the cloud providers' secret formatted to be a replaceable, or does the YAML explicitly look like this? Trying to determine if we have the correct style here. Thanks!

Comment (dgoodwin) [reply modules/admin-credentials-root-secret-formats.adoc:39]: An alternative would be using stringData instead of data, in which case the base64 encoding is not necessary. 

As written seems fine assuming people know that Base64encodeAccessKeyID means to base64 encode their access key ID.

Comment (joelddiaz) [reply modules/admin-credentials-root-secret-formats.adoc:39]: Yeah, stringData should be fine. It's when you 'oc get' the secret that you'll see the base64 representation.

**Comment by ahardin-rh** [comment]
/cherrypick enterprise-4.4

> Once this PR merges, I will create a separate PR for 4.3 to apply this content, but remove Mint + Remove Admin Cred from the AWS section per https://github.com/openshift/cloud-credential-operator#support-matrix

https://github.com/openshift/openshift-docs/pull/24381

**Pull request merged**
